src/platform/nrfconnect/wifi/WiFiManager.cpp

```diff
@@ -557,20 +511,7 @@ void WiFiManager::NetworkDrivenDisconnectHandler(Platform::UniquePtr<uint8_t>)
 
 void WiFiManager::ApplicationDrivenDisconnectHandler(Platform::UniquePtr<uint8_t>)
 {
-    if (!Instance().mRecoveryArmed)
```

Comment:
I think we should verify and better understand the behavior of the `NET_EVENT_WIFI_DISCONNECT_RESULT` and `NET_EVENT_WIFI_DISCONNECT_COMPLETE` events. Judging by the naming, the former just forwards if the disconnection request was successful and the latter indicates if the station was really disconnected from the AP. Yet, we used these two to distinguish between self-disconnects and disconnections forced by the application code with Instance().Disconnect(). The main reason why I raise this concern is that we need to know exactly when to inform the application that the station is disconnected with the following code:

```
    SystemLayer().ScheduleLambda([] {
        ChipLogProgress(DeviceLayer, "WiFi station disconnected");
        Instance().mWiFiState = WIFI_STATE_DISCONNECTED;
        Instance().PostConnectivityStatusChange(kConnectivity_Lost);
    });
```
which currently is placed in the `NetworkDrivenDisconnectHandler` but maybe it should be moved or copied here (depending on the relation between these two events).

Reply:
I can see that in the supplicant/driver code and the examples, they rely on `NET_EVENT_WIFI_DISCONNECT_RESULT` only. They use it to inform the upper layer that the connection has been closed. So in our implementation, after removing the workarounds, we don't need to know that the request comes from the application or network and we don't need to support both events anymore, we can leave only `NET_EVENT_WIFI_DISCONNECT_RESULT` and assign it to the disconnection callback. I will remove it properly.

Reply:
The `NET_EVENT_WIFI_DISCONNECT_COMPLETE` was added to ACK a disconnect request from the application saying that disconnection is now completed. `NET_EVENT_WIFI_DISCONNECT_RESULT` is a unsolicited notification that the connection no longer exists.

 

* App requests `NET_REQUEST_WIFI_DISCONNECT`
    - Once WPA supplicant completes disconnect it send `NET_EVENT_WIFI_DISCONNECT_COMPLETE`
* App requests connection and is success
    - AP disconnects the STA, then WPA supplicant sends `NET_EVENT_WIFI_DISCONNECT_RESULT`

We should probably rename `NET_EVENT_WIFI_DISCONNECT_RESULT` -> `NET_EVENT_WIFI_DISCONNECTED` or something.

Reply:
@krish2718 Thank you for the explanation! So @markaj-nordic, as I understood correctly, now, in our case, as we don't need to know if the Application wanted to disconnect or disconnection came from a different source, we should use only the NET_EVENT_WIFI_DISCONNECT_RESULT event to propagate it to the Application layer. Right?

Reply:
I'd say that we may want to keep both handlers because the `WiFiManager::Disconnect` is a part of public API and `NrfWiFiDriver` may call the  when reverting the configuration in the fail-safe mechanism. So in `NET_EVENT_WIFI_DISCONNECT_COMPLETE` handler we would simply have the code snippet from my first comment.

Reply:
Ok, so maybe we can just assign both events to the same callback, due to the fact that both of the callbacks will do exactly the same things?

Reply:
Sure, if the code is the same that would be perfect.